src/npc_attack.cpp

```diff
@@ -465,8 +466,9 @@ void npc_attack_gun::use( npc &source, const tripoint_bub_ms &location ) const
 
 bool npc_attack_gun::can_use( const npc &source ) const
 {
-    // can't attack with something you can't wield
-    return source.is_wielding( *gunmode ) || source.can_wield( *gunmode ).success();
+    // can't attack with something you can't wield or which lacks ammo.
+    return ( source.is_wielding( *gunmode ) || source.can_wield( *gunmode ).success() )
+           && gun.ammo_data() != nullptr && gun.ammo_data()->ammo != nullptr;
```

Comment:
I'm mildly concerned that `ammo_data()` could be silently expensive so this is now twice expensive, but I can't find an obvious proof of that.

Reply:
Introduced new operations for cheaper checks for presence of ammo and ammo_data and replaced ammo_data() calls where suitable.

I hope this addresses the concern of @akrieger, at the cost of some mild scope creep.

Reply:
Sorry I didn't mean to do all of that. I just know that other places, computing stuff like `ammo_data()` is expensive, so if you wanted to 'fix this' you'd just stash the return value in a local variable instead of inlining the function call into the return statement twice.

Reply:
Well, if it's a problem here it's a problem elsewhere, and so is better dealt with rather than skirted around with patches at random points.